flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/typeutils/LogicalRelDataTypeConverter.java

```diff
@@ -0,0 +1,649 @@
+            return relDataTypeFactory.createStructType(
+                    StructKind.PEEK_FIELDS_NO_EXPAND,
+                    rowType.getFields().stream()
+                            .map(f -> toRelDataType(f.getType(), relDataTypeFactory))
+                            .collect(Collectors.toList()),
+                    rowType.getFieldNames());
+        }
+
+        @Override
+        public RelDataType visit(DistinctType distinctType) {
+            throw new TableException("DISTINCT type is currently not supported.");
+        }
+
+        @Override
+        public RelDataType visit(StructuredType structuredType) {
+            final List<RelDataTypeField> fields = new ArrayList<>();
+            for (int i = 0; i < structuredType.getAttributes().size(); i++) {
+                final StructuredType.StructuredAttribute attribute =
+                        structuredType.getAttributes().get(i);
+                final RelDataTypeField field =
+                        new RelDataTypeFieldImpl(
+                                attribute.getName(),
+                                i,
+                                toRelDataType(attribute.getType(), relDataTypeFactory));
+                fields.add(field);
+            }
+            return new StructuredRelDataType(structuredType, fields);
+        }
+
+        @Override
+        public RelDataType visit(NullType nullType) {
+            return relDataTypeFactory.createSqlType(SqlTypeName.NULL);
+        }
+
+        @Override
+        public RelDataType visit(RawType<?> rawType) {
+            return new RawRelDataType(rawType);
+        }
+
+        @Override
+        public RelDataType visit(SymbolType<?> symbolType) {
+            return relDataTypeFactory.createSqlType(SqlTypeName.SYMBOL);
+        }
+
+        @Override
+        public RelDataType visit(LogicalType other) {
+            throw new TableException(
+                    String.format(
+                            "Logical type '%s' cannot be converted to a RelDataType.", other));
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // RelDataType to LogicalType
+    // --------------------------------------------------------------------------------------------
+
+    private static LogicalType toLogicalTypeNotNull(
+            RelDataType relDataType, DataTypeFactory dataTypeFactory) {
+        // dataTypeFactory is a preparation for catalog user-defined types
+        switch (relDataType.getSqlTypeName()) {
+            case BOOLEAN:
+                return new BooleanType(false);
+            case TINYINT:
+                return new TinyIntType(false);
+            case SMALLINT:
+                return new SmallIntType(false);
+            case INTEGER:
+                return new IntType(false);
+            case BIGINT:
+                return new BigIntType(false);
+            case DECIMAL:
+                if (relDataType.getScale() < 0) {
```

Comment:
What's the case that this code addresses? Forgive me if I didn't see, is it also tested?

Reply:
True, I added a test.